### DISCUSSION TITLE
Client certificate support for authentication

### DISCUSSION THREAD

**Discussion by ArsenArsen** [renovatebot/renovate#24926] [Suggest an Idea]
### What would you like Renovate to be able to do?

Client certificates are a useful authentication mechanism provided by TLS. They are used by some artifact servers, such as Artifactory, and `got` has support for loading PFX/PKCS12 certificates automatically.

### If you have any ideas on how this should be implemented, please tell us here.

#13991

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by d-g-n** [comment]
This issue is related to https://github.com/renovatebot/renovate/discussions/13896. I have checked out the branch from the PR and trialed it locally with a corporate artifactory behind certs and un/pw. Pointing the config to my company provided .p12 store works and allows self-hosted Renovate to work for our org, this would be a great addition.

**Comment by rarkins** [comment]
Please provide more details about both the requirements as well as your proposed solution, including links to external references

**Comment by d-g-n** [comment]
I can only speak about the requirements related to my use case, i can't comment on the implementation unfortunately.

We primarily use Scala/SBT and have several internal libraries that are used throughout our systems, these libraries are compiled and deployed to a central Artifactory (https://jfrog.com/artifactory/) that's custom configured to be only accessible when a developer or CI passes in both a dev cert in the form of a .p12 file, as well as a username and password.

In Renovate there is an ability to pass in a un/pw which works fine, but unfortunately because Renovate attempts to grab metadata from the org Artifactory, it fails and will not raise PRs specifically for our org dependencies, this makes Renovate a non-starter for us as keeping the link between public deps and library deps is essential.

In our current usage of SBT, it has the ability to communicate with this artifactory by passing in the cert path like so, `-Djavax.net.ssl.keyStore=cert.p12` alongside the `-Djavax.net.ssl.keyStorePassword=pass`. Having an equivalent of this in Renovate would be ideal. 

Hope that clears up the requirements, ideally this would also be able to apply generically to other non-standard repository setups in orgs that are behind certs.

**Comment by rarkins** [comment]
We also need to reference documentation for `got` - or perhaps it's Node.js itself - about how it support such certificates.

Another problem with the proposed solution is that it does not support repository config due to it reading from a file. Unless supporting repository is impossible or impractical, we would want the solution to support that.

**Comment by ArsenArsen** [comment]
> Another problem with the proposed solution is that it does not support repository config due to it reading from a file.

Why not? What would be required for that? **EDIT**: I misunderstood this as meaning archive repository/datasource config rather than Git repository.

https://github.com/sindresorhus/got/blob/9abdd06d72d90d034299dabd22544d14c897643b/documentation/5-https.md#pfx

**Comment by d-g-n** [comment]
> Another problem with the proposed solution is that it does not support repository config due to it reading from a file. Unless supporting repository is impossible or impractical, we would want the solution to support that.

I think given the fact that with the p12 alongside the passphrase are essentially keys to the kingdom for an org, even though i could see a feasible way for putting this in config by base64 encoding the file, then using the renovate encrypt utility which would allow it to work with the Renovate Github Application as well.

However, i don't think that's viable given the security concerns orgs would have with giving full access to anything the cert locks to a remote application, even with the fact it's encrypted. Maybe this is a moot concern given that the NPM encrypted token implementation i've seen elsewhere likely has a similar risk.

**Comment by viceice** [comment]
@ArsenArsen This option should be configurable like other hostRule settings, it should also be possible to configure it as string (pem encoded cert), so it can be configured in a repo `renovate.json`.  Checkout [`hostRules`](https://docs.renovatebot.com/configuration-options/#hostrules) for other options and search the code for them, so you know how to implement it.

**Comment by viceice** [comment]
Can be also similar to `privateKey` and `privateKeyPath`

- https://docs.renovatebot.com/self-hosted-configuration/#privatekey
- https://docs.renovatebot.com/self-hosted-configuration/#privatekeypath

**Comment by ArsenArsen** [comment]
I'd prefer

```ts
  keyStore?: {
    pkcs12: string | { filename: string; };
    passphrase?: string;
  };
```

or something along these lines. Keep in mind pkcs12 archives are binary data without a header or plaintext form. I can also provide alternatives for specifying the certificate and key manually but there's very little benefit, especially since that's a less used form of client side authentication. That'd look something like this:

```ts
  keyStore?: {
    pkcs12: string | { filename: string; };
    passphrase?: string;
  } | {
    certificate: string;
    key: string;
    passphrase?: string;
  };
```

**EDIT**: I'm unconvinced caching is a good idea here, it'd require restarting to update the cert (unless a file is `stat`ted before each access, which is worse than being reread for this filesize). Again, this is a <4kB file, it'll get cached by any worthwhile kernel, and reading it would require very little io time.

**Comment by rarkins** [comment]
How often is your p12 file changed? And does every developer currently need a copy in order to access artifactory?

**Comment by d-g-n** [comment]
> How often is your p12 file changed? And does every developer currently need a copy in order to access artifactory?

Every developer and tooling utility needs a generated .p12 to access artifactory yes.

In our org there's two main use cases:
- developer p12s are rotated once a year so using them with tooling isn't feasible
- we do have a CI tool built that acquires short-lived certs purely for tooling jobs, these expire within a few hours generally.

I believe there's a policy in place to not create infinite certificates as they can generally give very wide access to org systems. With that in mind, having the certs in the config to work with the Renovate Application would need further infrastructure on our side to auto rotate the certs in the repos.

That's not to say that this is the only way orgs do it, and i'm aware we're more on the careful side of the balance but it's not to say it's uncommon.

**Comment by rarkins** [comment]
So.. the p12 files are keys to the kingdom, and every developer has a key? I don't think it's logical to have security concerns about encrypting it into a bot config while no security concerns about having every developer have an unencrypted value of it permanently on their laptop, able to be read by the next computer virus or malicious npm package that comes around. And if you manage to rotate it once a year for developers, why is it not possible to rotate yearly in centralized tooling config? 

Do your developer certs *autorotate* somehow, and if so then what's that process?

Do you issue individual certificates for each developer, or a shared one?

What is the process for revoking certificates when a developer leaves the company?

Doubts about security aside, we are not saying that *you* must configure it in repository anyway. Instead, we are just saying we won't accept any solution which relies on reading it from disk only, and especially not per-request.

Also keep in mind that Renovate supports awaiting a `config.js` file so you could easily read the file value off disk there and then embed it into `hostRules`.

**Comment by ArsenArsen** [comment]
> Also keep in mind that Renovate supports awaiting a `config.js` file so you could easily read the file value off disk there and then embed it into `hostRules`.

I was going to ask about this, actually, but I forgot to. Since that's the case, `keyStore?: { pkcs12: string | buffer; passphrase?: string | buffer; }` is probably the best option. pkcs12 being a b64 string or a raw pkcs12 buffer, and passphrase being either plaintext (no b64) or any arbitrary buffer. Thoughts?

**Comment by ArsenArsen** [comment]
Also, again, this antagonizing of reading a <4kB file once per request is ridiculous.

```js
> console.time(); for (let i = 0; i < 1_000_000; i++) { fs.readFileSync("/home/arsen/client-identity.p12"); }; console.timeEnd()
default: 7.405s
undefined
> // edit to add fs/promises equivalent:
> const fsPromises = require("fs/promises")
undefined
> fsPromises.readFile
[AsyncFunction: readFile]
> console.time(); for (let i = 0; i < 1_000_000; i++) { await fsPromises.readFile("/home/arsen/client-identity.p12"); }; console.timeEnd()
default: 3:25.194 (m:ss.mmm)
undefined
>
```
